mlflow/spacy.py

```diff
@@ -295,3 +299,114 @@ def load_model(model_uri):
+                   autologging.
+    :param run_name: If provided, the run is started with the given name.
+    :param remove_config_values: Parameters with these keys are removed from logging
+    :param log_models: If ``True``, trained models are logged as MLflow model artifacts.
+                       If ``False``, trained models are not logged.
+    :param log_interval: When provided, the pipeline object be logged into checkpoints directory as 
+                        MLFlow models artifacts and the metrices will be logged at the given intervals.
+    :param log_dataset_dir: When the path to the corpus/dataset directory is given, the dataset will
+                            be logged into datasets directory as an artifact.
+    :param log_best_model: If ``True``, the spaCy pipeline with maximum value for matric ``score`` 
+                            will be logged into best-model directory as MLFlow spacy model.
+    """
+
+    import spacy
+    import re
+
+    def setup_logger(
+        nlp: spacy.language.Language, stdout: IO = sys.stdout, stderr: IO = sys.stderr
+    ) -> Tuple[Callable[[Dict[str, Any]], None], Callable[[], None]]:
+        mlflow.start_run(run_name=run_name)
+        config = nlp.config.interpolate()
+        config_dot = spacy.util.dict_to_dot(config)
+        for _field in remove_config_values:
+            del config_dot[_field]
+        pattern = re.compile('[\W]+')
+        mlflow.log_params(
+            {pattern.sub('.', _field): _value for _field, _value in config_dot.items()})
+
+        def _log_metrics(metric: Dict[str, Any], parent: str = '', step: Optional[int] = None):
+            for k, v in metric.items():
+                if isinstance(v, (dict,)):
+                    if parent:
+                        _log_metrics(v, f"{parent}.{k}", step)
+                    else:
+                        _log_metrics(v, k, step)
+                else:
+                    if parent:
+                        mlflow.log_metric(f"{parent}.{k}", v, step=step)
+                    else:
+                        mlflow.log_metric(k, v, step=step)
+
+        if log_dataset_dir:
+            mlflow.log_artifact(log_dataset_dir, "dataset")
+
+        def log_step(info: Optional[Dict[str, Any]]):
+            if info is not None and log_interval:
+                if info["step"] % log_interval == 0 and info["step"] != 0:
+                    _log_metrics({k: info.get(k) for k in (
+                        "losses", "score", "other_scores")}, step=info.get("step"))
+                    if log_models:
+                        log_model(
+                            nlp, f"models/checkpoints/epoch_{info['epoch']}_step_{info['step']}")
+                        if log_best_model and info["score"] == max(info["checkpoints"])[0]:
+                            log_model(nlp, f"models/best-model")
+
+        def finalize() -> None:
+            mlflow.end_run()
+
+        return log_step, finalize
+
+    return setup_logger
```

Comment:
It feels like in this implementation we are returning the method for spaCy to use, while in other autolog implementations they're safe_patching:
https://github.com/mlflow/mlflow/blob/0f9d1e04975c69313ad57e4d7d428ab7befad69e/mlflow/utils/autologging_utils/safety.py#L272-L277

See gluon:

https://github.com/mlflow/mlflow/blob/0f9d1e04975c69313ad57e4d7d428ab7befad69e/mlflow/gluon.py#L427

Or lightBGM:

https://github.com/mlflow/mlflow/blob/0f9d1e04975c69313ad57e4d7d428ab7befad69e/mlflow/lightgbm.py#L563-L564

Or xgboost:
https://github.com/mlflow/mlflow/blob/0f9d1e04975c69313ad57e4d7d428ab7befad69e/mlflow/xgboost.py#L672-L673

for reference

Reply:
Thanks for the suggestion, I think I'll either have to rename the functionality to something other than autologging, or refactor the architecture to something similar to all of these. It would be really helpful to get some suggestions from community on this